### DISCUSSION TITLE
cocmahua

### DISCUSSION THREAD

**Discussion by zhongfly** [zhongfly/tachiyomi-extensions#36] [General]
@lmk1988 Sorry to bother you.Since there is no other way to contact you, this is the only way to communicate with you.
You used to be the main maintainer of the cocomanga plugin in tachiyomi.
But https://github.com/tachiyomiorg/tachiyomi-extensions/pull/11365 mentions that the cocomanga plugin will be removed if no one continues to maintain it. Want to ask if you will continue to maintain it? cocomanga is a good source for manga, I would like to have the opportunity to continue using it if possible

**Comment by lmk1988** [comment]
Opps, just saw this, but yea just submitted MR to update the encryption keys.

Very interesting new protection mechanism. Would love to know how they are doing the new obfuscation.

But I wouldn't be able to update it immediately when they patch their protection, saw that they plan to remove the plugin if the site do any updates again.

**Reply by zhongfly** [reply]
Thanks for the update, the site may not update encryption frequently, the last time was January 2021.

**Reply by zhongfly** [reply]
I'm also looking to learn how to crack the encryption for this site. I read https://www.cnblogs.com/Amd794/p/14960487.html, but it is a bit outdated, I found that the website has added anti-debugging, and the related code has been completely changed. I wonder if you are interested in writing an article documenting the reverse process?

**Reply by lmk1988** [reply]
Will be too tedious to write an article. Plus I think they are trying to prevent other sites from scrapping them instead, so it wouldn't make sense to publish the article.

The flow did not change (since it is harder to update flow).

Just the keys and obfuscation engine changed.

Since there's decryption involved. You just need to overwrite the decryption function (via proxy response body overwrite) and overwrite it to print out the keys to some other global variable.

But the tedious process would be to identify where the new decryption method is at

**Comment by zhongfly** [comment]
@lmk1988 https://github.com/tachiyomiorg/tachiyomi-extensions/pull/11365#issuecomment-1096074968 
The website has changed the encryption again, I have a proposal: directly abandon the submission of pr to the compelling tachiyomi-extensions repository, but only update the code in the personal repository

**Reply by lmk1988** [reply]
I guess the extension was heavily used, now the traffic will flow to other sites instead.

Not sure why they want to remove it so fast though. There're other plugins that are breaking, but just not reported.

Will see if the personal repo makes sense or just find some other sites to move to

**Reply by lmk1988** [reply]
But it seems to be still working for me though, is it no longer working for you?

**Reply by zhongfly** [reply]
> Will see if the personal repo makes sense or just find some other sites to move to

For me, cocomanga is irreplaceable, it uploads Bilibili's translation comics, and some of the comics are not available on other sites that also upload Bilibili comics (for example, copmag does not have "古见同学有交流障碍症")

**Reply by zhongfly** [reply]
> But it seems to be still working for me though, is it no longer working for you?

When I open the comic chapter to read, it is prompted that the decryption failed(encodedData exhausted keys)

**Reply by lmk1988** [reply]
Will update it privately without showing the new keys, see if they may stop changing it.

https://github.com/lmk1988/tachiyomi-extensions/releases

**Reply by zhongfly** [reply]
Can I post release address in tachoyomi's Chinese chat group (unofficial, created by chinese fans themselves)?

**Reply by lmk1988** [reply]
Is it possible to only share the apk?

**Reply by zhongfly** [reply]
> Is it possible to only share the apk?

ok

**Reply by zhongfly** [reply]
![downloads](https://badgen.net/github/assets-dl/lmk1988/tachiyomi-extensions/v1.2.12-private?cache=300)

you can use this badge to to know how many times the release has been downloaded
```
![downloads](https://badgen.net/github/assets-dl/lmk1988/tachiyomi-extensions/v1.2.12-private?cache=300)
```

**Reply by zhongfly** [reply]
I got a suggestion, which I think might ease your load: fill in the change key in the settings. This way, when the key changes, just distribute the key

**Reply by lmk1988** [reply]
Currently have found a way to auto crack the keys, but I seem to be blocked by cloudflare too often to test it out. Hopefully can resolve it these 2 weeks

**Reply by lmk1988** [reply]
Ok, hope this release is better

https://github.com/lmk1988/tachiyomi-extensions/releases/tag/v1.2.14-private

**Reply by zhongfly** [reply]
cocomanga strengthens the cloudflare crawler verification again…

**Reply by lmk1988** [reply]
It's not 100% working, but I think it's just a matter of time they reduce the protection. They're chasing away their user base with this strengthen protection as well.

I think the current protection using tachiyomi's cloudflare client can be tweaked further, perhaps it's just the waiting time is too short now, may need a custom http client to handle this behaviour I think

**Reply by stevenyomi** [reply]
@lmk1988 如果不再更新的话，可以分享一下修改后的源码吗？

**Reply by lmk1988** [reply]
Added to private repo

**Comment by zhongfly** [comment]
Maybe consider using this site https://manga8.xyz/ , I found they stole manga from cocomanga site

**Reply by lmk1988** [reply]
Will see if I can reuse coco codes for this. Coco keep changing their codes, I guess need wait quite awhile for them to stabilize

**Reply by zhongfly** [reply]
I found that this website is not as complicated as coco.
```html
<script> var oScript = document.createElement('script'); oScript.src = "https://img.manga8.xyz/cocomh/txt/1236/352.txt"; document.body.appendChild(oScript); function show(data) { document.getElementById('content_neirong').innerHTML = data; } </script>
```
https://img.manga8.xyz/cocomh/txt/1236/352.txt records the manga pictures.
```js
show('<p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/1.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/2.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/3.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/4.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/5.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/6.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/7.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/8.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/9.jpg"></p><p><img src="http://img.manga8.xyz/cocomh/tupian/1236/352/10.jpg"></p>')
```

**Reply by lmk1988** [reply]
Added [MR](https://github.com/tachiyomiorg/tachiyomi-extensions/pull/11690)

**Reply by zhongfly** [reply]
> Added [MR](https://github.com/tachiyomiorg/tachiyomi-extensions/pull/11690)

I foud https://www.wt0.net/ https://www.darpou.com/ are same as https://manga8.xyz/.

**Reply by zhongfly** [reply]
> Added [MR](https://github.com/tachiyomiorg/tachiyomi-extensions/pull/11690)

The order of chapters in baimangu needs to be reversed, currently older chapters are considered newest, and newest chapters are considered first.

**Comment by lmk1988** [comment]
@stevenyomi 

So Cocomanhua seems to have lowered their cloudfront protection finally.
But they have upgraded their image protection, there's a need to decrypt the images now.

1. Decrypt `C_DATA` to get `mh_info` and `image_info`
2. Decrypt `mh_info.enc_code1` to get total number of pages
3. Decrypt `mh_info.enc_code2` to get the image path `22934/aHhtdWQ5S3BycXRxb2xTL0xsdWEyekFoVWlCUk5KMWJvNWQ5QTFCYXVnST0=/`
4. Prefix image path with `https://img.cocomanga.com/comic/`
5. Suffix image path with `0001.enc.webp`, running numbers start from 1 to the "total number of pages"

```
https://img.cocomanga.com/comic/22934/aHhtdWQ5S3BycXRxb2xTL0xsdWEyekFoVWlCUk5KMWJvNWQ5QTFCYXVnST0=/0001.enc.webp
```

So that was the previous way to do things.

--------------------

Additional steps to do now:
1. Decrypt `image_info.imgKey` to get imageDecryptionKey
2. Read the image content and decrypt the image content with the imageDecryptionKey
```
let data = convertUint8ArrayToWordArray(new Uint8Array(await file.arrayBuffer()));
let cryptoParam = {
  ciphertext: data,
};
let decryptedVal = CryptoJS.AES.decrypt(cryptoParam, CryptoJS.enc.Utf8.parse(decryptKey), {
  iv: CryptoJS.enc.Utf8.parse('0000000000000000'),
  mode: CryptoJS.mode.CBC,
  padding: CryptoJS.pad.Pkcs7,
});
let imageData = new Blob([convertWordArrayToUint8Array(decryptedVal)['buffer']]);
img.src = URL.createObjectURL(imageData);
```
Something like this ^

Then the image data should be decrypted and ready for render


----------------------
Sample HTML code here for decrypting cocomanhua images
```
<head>
  <title>Testing decode cocomanga image</title>
  <script
    type="text/javascript"
    src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"
  ></script>
  <script
    src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/4.1.1/crypto-js.min.js"
    integrity="sha512-E8QSvWZ0eCLGk4km3hxSsNmGWbLtSCSUcewDQPQWZF6pEU8GlT8a5fF32wOl1i8ftdMhssTrF/OhyGWwonTcXA=="
    crossorigin="anonymous"
    referrerpolicy="no-referrer"
  ></script>
</head>
<body>
  <input type="file" id="file" /><br />
  <img id="img" />
</body>
<script>
  convertWordArrayToUint8Array = function (wordArray) {
    var len = wordArray.words.length,
      u8_array = new Uint8Array(len << 2),
      offset = 0,
      word,
      i;
    for (i = 0; i < len; i++) {
      word = wordArray.words[i];
      u8_array[offset++] = word >> 24;
      u8_array[offset++] = (word >> 16) & 0xff;
      u8_array[offset++] = (word >> 8) & 0xff;
      u8_array[offset++] = word & 0xff;
    }
    return u8_array;
  };
  convertUint8ArrayToWordArray = function (u8Array) {
    var words = [],
      i = 0,
      len = u8Array.length;
    while (i < len) {
      words.push(
        (u8Array[i++] << 24) |
          (u8Array[i++] << 16) |
          (u8Array[i++] << 8) |
          u8Array[i++]
      );
    }
    return {
      sigBytes: words.length * 4,
      words: words,
    };
  };
</script>
<script>
  const decryptKey = '6Cj7a6euGfWeQxwh';
  decodeImageFile = async function(file) {
    let data = convertUint8ArrayToWordArray(new Uint8Array(await file.arrayBuffer()));
    let cryptoParam = {
      ciphertext: data,
    };
    let decryptedVal = CryptoJS.AES.decrypt(cryptoParam, CryptoJS.enc.Utf8.parse(decryptKey), {
      iv: CryptoJS.enc.Utf8.parse('0000000000000000'),
      mode: CryptoJS.mode.CBC,
      padding: CryptoJS.pad.Pkcs7,
    });
    return new Blob([convertWordArrayToUint8Array(decryptedVal)['buffer']]);
  }
</script>
<script>
  var urlObject = window.URL || window.webkitURL;
  $("#file").change(async function (e) {
    var file;
    if ((file = this.files[0])) {
      // setup image element
      img = document.getElementById("img");
      img.onerror = function () {
        alert("image error");
      };
      let imageData = await decodeImageFile(file);
      img.src = URL.createObjectURL(imageData);
    }
  });
</script>

```
